bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/auth/client/oauth2/OAuthFactory.java

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.core.auth.client.oauth2;
+
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.Nullable;
+
+/**
+ * The OAuth Factory interface
+ *
+ * @author Michael Bock - Initial Contribution
+ * @author Gary Tse - ESH Adaptation
+ */
+public interface OAuthFactory {
+
+    /**
+     * Creates a new oauth service. Use this method only once to obtain a handle and store
+     * this handle for further in a persistent storage container.
+     *
+     * @param tokenUrl the token url of the oauth provider. This is used for getting access token.
+     * @param authorizationUrl the authorization url of the oauth provider. This is used purely for generating
+     *            authorization code/ url.
+     * @param clientId the client id
+     * @param clientSecret the client secret (optional)
+     * @param scope the desired scope
+     * @param supportsBasicAuth whether the service supports basic authorization or the client id and client secret
+     *            should be
+     *            passed as form params
+     * @return a handle to the oauth service
+     */
+    String createOAuthClientService(@NonNull String tokenUrl, @Nullable String authorizationUrl,
+            @NonNull String clientId, @Nullable String clientSecret, @Nullable String scope,
+            @Nullable Boolean supportsBasicAuth);
+
+    /**
+     * Gets the oauth service for a given handle
+     *
+     * @param handle the handle to the oauth service
+     * @return the oauth service, or null of the handle does not exist
+     */
+    OAuthClientService getOAuthClientService(@NonNull String handle);
+
+    /**
+     * Unget an oauth service, this unget/unregister the service, and frees the resources.
+     * The existing tokens/ configurations (persisted parameters) are still saved
+     * in the store.
+     *
+     * @param handle the handle to the oauth service
+     */
+    void ungetOAuthService(@NonNull String handle);
+
+    /**
+     * This method is for unget/unregister the service,
+     * then <strong>DELETE</strong> access token, configuration data from the store
+     *
+     * @param handle
+     */
+    void deleteServiceAndAccessToken(@NonNull String handle);
+
+    /**
+     * This unget/register all services, and free the resources.
+     * After that all presisted AccessTokens, configuration data are
+     * also <strong>DELETED</strong> from the store.
+     *
+     * <strong>Use with extreme caution!!!</strong>
```

Comment:
When I wrote this, I think there will be data privacy needs to remove all tokens.  The problem is that the tokens are persisted in the store and cant be removed without the handle.  

Even uninstall of the oauth implementation does not remove tokens.  

I think there's pros & cons for both sides.  What's your verdict?

Reply:
It should imho be removed.
I do not see any general ESH use case, where you would call this. We currently do not have any feature like "wipe all user-specific information".

So unless you have a requirement (today) for this method, leave it out - it can be easily added at the moment when we notice that we need it.

Reply:
Removed in next commit